lib/src/rules/unreachable_from_main.dart

```diff
@@ -142,12 +141,30 @@ class _ReferenceVisitor extends RecursiveAstVisitor {
     super.visitAnnotation(node);
   }
 
+  @override
+  void visitAsExpression(AsExpression node) {
+    node.expression.accept(this);
+    // Intentionally do not visit `node.type`, as a reference in the type of an
+    // as-expression is not good enough to count as "reachable". Marking a type
+    // as reachable only because it was seen in an as-expression would be a
+    // miscategorization if the type is never instantiated or subtyped.
```

Comment:
Just curious, but does the same reasoning apply for type annotations?
```
class _C {}
void f() {
  _C? c; // Is this a use? It's certainly dead code.
}
```

Reply:
I _think_ that type annotations should also not count as a "use". I don't think we have that logic for `unused_element` so I decided not to include it in this PR. But I think a type annotation would be the same as an as-expression, etc (in fact, its typically the same AST node, TypeAnnotation). I want to think about parameter type annotations, local, type arguments, type parameter bounds etc.... (maybe none of them count as a use...)